app-shells/autojump: add FISH support, misc. clean ups

marcecj opened the pull request:
This goal of this PR is to add FISH support.  This is done in the first commit, which required including two upstream patches (they can be removed with the next version bump, though).  The other three commits are pretty much just clean ups that I made while working on FISH support, so decided to include them here.

Note that I made FISH support unconditional, since zsh support is also unconditional, and it only installs a single file.

I'm not sure if the second patch is really required, but it seems like a good idea to me.  IIUIC, it will limit the autojump script to the python implementations listed in `PYTHON_COMPAT`, that is, it will cause the script to fail for unsupported Python implementations.

For easier comparison, here's a diff of the ebuild revisions:

    % git diff master.. -C -C ./*.ebuild
    diff --git a/app-shells/autojump/autojump-22.2.4-r3.ebuild b/app-shells/autojump/autojump-22.2.4-r4.ebuild
    similarity index 55%
    copy from app-shells/autojump/autojump-22.2.4-r3.ebuild
    copy to app-shells/autojump/autojump-22.2.4-r4.ebuild
    index 4b774f9..6c967ce 100644
    --- a/app-shells/autojump/autojump-22.2.4-r3.ebuild
    +++ b/app-shells/autojump/autojump-22.2.4-r4.ebuild
    @@ -26,6 +26,15 @@ src_prepare() {
     	sed -e "s: \(/etc/profile.d\): \"${EPREFIX}\1\":" \
     		-e "s:/usr/local/share:/usr/share:" \
     		-i bin/autojump.sh || die
    +
    +	# autojump_argparse is only there for Python 2.6 compatibility
    +	sed -e "s:autojump_argparse:argparse:" \
    +		-i bin/autojump || die
    +
    +	# upstream fixes to the autojump.fish script; the first patch is needed for
    +	# the second patch to apply
    +	epatch "${FILESDIR}/autojump-22.4.4-fix-autojump.fish-bugs.patch"
    +	epatch "${FILESDIR}/autojump-22.4.4-fix-__aj_error-typo.patch"
     }
     
     src_compile() {
    @@ -34,6 +43,7 @@ src_compile() {
     
     src_install() {
     	dobin bin/autojump
    +	python_replicate_script "${ED}"/usr/bin/autojump
     
     	insinto /etc/profile.d
     	doins bin/"${PN}".sh
    @@ -41,16 +51,29 @@ src_install() {
     	insinto /usr/share/"${PN}"/
     	doins bin/"${PN}.bash"
     	doins bin/"${PN}.zsh"
    +	doins bin/"${PN}.fish"
     	insinto /usr/share/zsh/site-functions
     	doins bin/_j
     
    -	python_foreach_impl python_domodule bin/autojump_argparse.py bin/autojump_data.py bin/autojump_utils.py
    +	python_foreach_impl python_domodule bin/autojump_data.py bin/autojump_utils.py
     	if use python; then
     		python_foreach_impl python_domodule tools/autojump_ipython.py
    -		einfo 'This tool provides "j" for ipython, please add'
    -		einfo '"import autojump_ipython" to your ipy_user_conf.py.'
     	fi
     
     	doman docs/"${PN}.1"
     	dodoc README.md
     }
    +
    +pkg_postinst() {
    +	if use python; then
    +		elog 'This tool provides "j" for ipython, please add'
    +		elog '"import autojump_ipython" to your ipy_user_conf.py.'
    +		elog
    +	fi
    +
    +	elog 'If you use app-shells/fish, add the following code to your'
    +	elog 'config.fish to get autojump support:'
    +	elog 'if test -f /usr/share/autojump/autojump.fish'
    +	elog '    source /usr/share/autojump/autojump.fish'
    +	elog 'end'
    +}

marcecj commented:
There's one more change I would like to make, but am not sure about: the `python` USE flag is used to control support for `dev-python/ipython`, so I was going to rename it to `ipython`, as used in four other packages.  Does that sound OK?

marcecj commented:
OK, wow, that was a particularly bad commit to base my branch off of!  I don't see anything pertaining to autojump in the CI output, though.

marcecj commented:
@xmw @gentoo/reviewers

marcecj commented:
OK, @gentoo/reviewers doesn't seem to work, so I'll try @mgorny instead.

monsieurp commented:
@marcecj yo. can you commit the change you wanted to check in about python so that the CI tests run again? thanks.

marcecj commented:
OK, I made the changes.  I added a commit and rebased off of current master.  Here's the new diff:

    % diff -u autojump-22.2.4-r*.ebuild
    --- autojump-22.2.4-r3.ebuild   2015-11-29 09:54:50.029586746 +0100
    +++ autojump-22.2.4-r4.ebuild   2015-12-04 22:00:21.457486239 +0100
    @@ -15,17 +15,26 @@
     LICENSE="GPL-3"
     SLOT="0"
     KEYWORDS="~amd64 ~x86 ~ppc ~ppc64"
    -IUSE="python test"
    +IUSE="ipython test"
     
     # Not all tests pass. Need investigation.
     RESTRICT="test"
    -RDEPEND="python? ( ${PYTHON_DEPS} )"
    +RDEPEND="ipython? ( ${PYTHON_DEPS} )"
     DEPEND="test? ( dev-python/flake8 dev-python/tox )"
     
     src_prepare() {
            sed -e "s: \(/etc/profile.d\): \"${EPREFIX}\1\":" \
                    -e "s:/usr/local/share:/usr/share:" \
                    -i bin/autojump.sh || die
    +
    +       # autojump_argparse is only there for Python 2.6 compatibility
    +       sed -e "s:autojump_argparse:argparse:" \
    +               -i bin/autojump || die
    +
    +       # upstream fixes to the autojump.fish script; the first patch is needed for
    +       # the second patch to apply
    +       epatch "${FILESDIR}/autojump-22.4.4-fix-autojump.fish-bugs.patch"
    +       epatch "${FILESDIR}/autojump-22.4.4-fix-__aj_error-typo.patch"
     }
     
     src_compile() {
    @@ -34,6 +43,7 @@
     
     src_install() {
            dobin bin/autojump
    +       python_replicate_script "${ED}"/usr/bin/autojump
     
            insinto /etc/profile.d
            doins bin/"${PN}".sh
    @@ -41,16 +51,29 @@
            insinto /usr/share/"${PN}"/
            doins bin/"${PN}.bash"
            doins bin/"${PN}.zsh"
    +       doins bin/"${PN}.fish"
            insinto /usr/share/zsh/site-functions
            doins bin/_j
     
    -       python_foreach_impl python_domodule bin/autojump_argparse.py bin/autojump_data.py bin/autojump_utils.py
    -       if use python; then
    +       python_foreach_impl python_domodule bin/autojump_data.py bin/autojump_utils.py
    +       if use ipython; then
                    python_foreach_impl python_domodule tools/autojump_ipython.py
    -               einfo 'This tool provides "j" for ipython, please add'
    -               einfo '"import autojump_ipython" to your ipy_user_conf.py.'
            fi
     
            doman docs/"${PN}.1"
            dodoc README.md
     }
    +
    +pkg_postinst() {
    +       if use ipython; then
    +               elog 'This tool provides "j" for ipython, please add'
    +               elog '"import autojump_ipython" to your ipy_user_conf.py.'
    +               elog
    +       fi
    +
    +       elog 'If you use app-shells/fish, add the following code to your'
    +       elog 'config.fish to get autojump support:'
    +       elog 'if test -f /usr/share/autojump/autojump.fish'
    +       elog '    source /usr/share/autojump/autojump.fish'
    +       elog 'end'
    +}

monsieurp commented:
It looks good. I will merge it now. Thanks a lot btw for the prompt answer.

monsieurp commented:
I sent him an email yesterday. Let's wait.
